### PR TITLE
nixos/doc: suggest checking options for declarative package mgmt

### LINKED ISSUE #77959: TODO: mention nixos-option and nixos.org/nixos/options.html in manual Chapter 6
We should mention in nixos manual Chapter 6 that options can be dicovered with `nix-option` and http://nixos.org/nixos/options.html. It will help newbs learn how to discover packages more easily.

As a new user, I was sticking everything I wanted inside of the `systemPackages` array, and I didn't not realize that `programs.foo.enable` is a better route to take because it configures many things automatically. For example, `programs.zsh.enable = true` results in `nix-zsh-completions` also being installed as well as features enabled, while writing `environment.systemPackages = [pkgs.zsh]` does not.

This will be helpful to new users.

### PR DESCRIPTION
**Pull request by jtojnar**
Some programs might not work without systemd services registered or do not work at full capacity without NixOS module.

Closes: https://github.com/NixOS/nixpkgs/issues/77959

### CODE REVIEW

**Comment by jtojnar** [comment]
Rendered:

image

**Comment by trusktr** [comment]
Nice. I think it may be nice to have wording with `nixos-option` and a link to `options.html`. Where does `ch-options` link to? <sub>The docs system is a bit hard to navigate.</sub>

**Comment by jtojnar** [comment]
`#ch-options` links to the Options chapter, which, on the web, is located at `options.html`.

**Comment by jtojnar** [comment]
IMHO, `nixos-option` is not very useful for discovery unless you already roughly know what option you want – not all packages are programs.

**Comment by trusktr** [comment]
So `#ch-options` links to https://nixos.org/nixos/manual/options.html?

@jonringer's quote [from the forums](https://discourse.nixos.org/t/programs-foo-enable-true-vs-systempackages-foo-is-confusing/5534/9) is good. Here's an updated version with more detail

---

Generally it is recommended to use options (search available options at [nixos.org/nixos/options.html](https://nixos.org/nixos/options.html)) instead of `environment.systemPackages` for configuring which packages are installed. Options are smart about system effects needed for applications/services to run correctly.

For example,
- setting `virtualization.docker.enable = true` will create the necessary groups, install the docker-daemon, install the docker cli, and a few other things to enable normal docker usage.
- setting `programs.zsh.enable = true` will install `zsh` along with `nix-zsh-completions` as well as enable completions and other features by default, whereas installing `zsh` by writing `environment.systemPackages = [pkg.zsh]` will install only `zsh` without `nix-zsh-completions` and other features.

Not all packages have a `programs` (or similar) option and in those cases installing them via `environment.systemPackages` is the only way.

---

**Comment by trusktr** [comment]
@jtojnar I edited the above. Can you adapt that? I think it is better as note a "note" section.

**Comment by jtojnar** [comment]
Not really sure if this would not derail [the section](https://nixos.org/nixos/manual/index.html#sec-declarative-package-mgmt) too much. Especially for something that is not relevant to the majority of packages.

**Comment by flokli** [comment]
I also think @trusktr's message is somewhat misleading - most of the time, just installing the package is just fine, sometimes it needs some custom dbus or udev magic, for which modules exist, but we don't have modules for all packages in nixpkgs.

I'd propose to merge @jtojnar 's version for now (as it already improves the situation quite a lot!), and if we come up with something better, do that in a follow-up PR.

**Comment by worldofpeace** [comment]
Yay, thanks so much @jtojnar :sparkle: . This is an extremely important part of documentation that was missing.

**Comment by Raikiri** [comment]
Correct me if I'm wrong, but this page as of now is still missing any mention that packages can (and when possible, should?) be installed as modules: https://nixos.org/manual/nixos/stable/index.html#sec-package-management

**Comment by flokli** [comment]
The note introduced in this PR did end up in the docs:

image


Do you have a proposal on how to better word this?

**Comment by Raikiri** [comment]
The way I read the current note is like this: "Some packages are hard to install and you will fail if you attempt to follow the guide below. So in order to install them, you'll need to play with their options by following this guide".

The wording is very unclear that "options" actually refers to an altogether different way of installing software (via modules) and it does not mention that the same package can be often installed as a package or as a module (that does the same thing + configuration) and that in fact modules should be preferred (?) when possible. 

I'm not knowledgeable enough about the system to suggest how it should be worded, so I'm only reporting from a user's point of view, but in my opinion the way it should be formatted is something like this:

In addition to installing packages by adding them to the list of system packages, some of them can be installed using a mechanism of options/modules:
`programs.steam.enable=true;`
This way should be preferred when available because in addition to installing packages it also configures the environment, here's how you search for packages that are available via modules and here's a link of where to read more on more things that are available via options.

TL;DR: currently it's very unclear that "options" actually refers to an alternative (and actually preferred) way of installing software.

**Comment by flokli** [comment]
cc @roberth
